src/aleph/web/controllers/storage.py

```diff
@@ -56,20 +80,133 @@ async def add_storage_json_controller(request: web.Request):
+    actual_item_hash = sha256(content).hexdigest()
+    c_item_hash = item_content["item_hash"]
+
+    is_signature = await verify_signature(message=pending_message_db)
+    with session_factory() as session:
+        current_balance = get_total_balance(
+            session=session, address=pending_message_db.sender
+        )
+    if current_balance < len(content):
+        output = {"status": "Payment Required"}
+        return web.json_response(output, status=402)
+    if not is_signature:
+        output = {"status": "Forbidden"}
+        return web.json_response(output, status=403)
+    elif actual_item_hash != c_item_hash:
+        output = {"status": "Unprocessable Content"}
+        return web.json_response(output, status=422)
+    elif len(content) > 25_000 and not message:
+        output = {"status": "Unauthorized"}
+        return web.json_response(output, status=401)
+    else:
+        return None
+
+
+async def storage_add_file_with_message(request: web.Request):
     storage_service = get_storage_service_from_request(request)
     session_factory = get_session_factory_from_request(request)
+    config = get_config_from_request(request)
 
-    # No need to pin it here anymore.
-    # TODO: find a way to specify linked ipfs hashes in posts/aggr.
     post = await request.post()
     file_io = multidict_proxy_to_io(post)
+    message, size = await get_message_content(post)
+    pending_message_db = PendingMessageDb.from_message_dict(
+        message_dict=message, reception_time=dt.datetime.now(), fetched=True
+    )
+    mq_channel = await get_mq_channel_from_request(request, logger=logger)
+    mq_queue = await mq_make_aleph_message_topic_queue(
+        channel=mq_channel,
+        config=config,
+        routing_key=f"*.{pending_message_db.item_hash}",
+    )
+
+    is_valid_message = await verify_and_handle_request(
+        pending_message_db, file_io, message, size, session_factory
+    )
+    if is_valid_message is not None:
+        return is_valid_message
+
     with session_factory() as session:
         file_hash = await storage_service.add_file(
             session=session, fileobject=file_io, engine=ItemType.storage
         )
+        session.add(pending_message_db)
         session.commit()
+    mq_message = await _mq_read_one_message(mq_queue, 30)
+
+    if mq_message is None:
+        output = {"status": "accepted"}
+        return web.json_response(output, status=202)
+    if mq_message.routing_key is not None:
+        status_str, _item_hash = mq_message.routing_key.split(".")
+        processing_status = MessageProcessingStatus(status_str)
+        status_code = _processing_status_to_http_status(processing_status)
+        return web.json_response(status=status_code, text=file_hash)
+
+
+async def storage_add_file(request: web.Request):
+    post = await request.post()
+    if post.get("message", b"") is not None and post.get("size") is not None:
+        return await storage_add_file_with_message(request)
```

Comment:
This split is not ideal because you duplicate the part where you read the file and add it in the DB. Also, there is a major bug in your new function because you never actually write the file to disk nor add the corresponding entry in the `files` table. I'd rather have one function where you call other functions, example (fleshed out, just to give an example):

```
if message and size:
  pending_message = await verify_message(message)
  await verify_balance(pending_message.content.address, size)
  expected_file_hash = pending_message.content.item_hash  
else:
  expected_file_hash = None

file_io = multidict_proxy_to_io(post)
file_content, item_hash = await read_and_check_file(file_io, expected_file_hash)

# Note that this requires overloading the `StorageService.add_file` method to support passing bytes instead of an IO object 
await storage_service.add_file(session=session, file=file_content, engine=ItemType.storage)
```